decode.go

```diff
@@ -137,30 +139,59 @@ func Uncompress(b []byte) (dst []byte) {
 
 // Decodes data from src in q ipc format.
 func Decode(src *bufio.Reader) (data *K, msgtype int, e error) {
-	var header ipcHeader
-	e = binary.Read(src, binary.LittleEndian, &header)
+	header, e := DecodeHeader(src)
 	if e != nil {
 		return nil, -1, errors.New("Failed to read message header:" + e.Error())
 	}
+	// try to buffer entire message in one go
+	src.Peek(int(header.MsgSize - 8))
+	return DecodeData(src, header)
+}
+
+func DecodeHeader(src *bufio.Reader) (*ipcHeader, error) {
+	var header ipcHeader
+	e := binary.Read(src, binary.LittleEndian, &header)
+	if e != nil {
+		return nil, e
+	}
 	if !header.ok() {
-		return nil, -1, errors.New("header is invalid")
+		return nil, errors.New("header is invalid")
+	}
+	return &header, nil
+}
+
+func DecodeRaw(src *bufio.Reader) ([]byte, *ipcHeader, error) {
```

Comment:
Specifically, I used it to bypass all parsing (apart from the header) and have the code pass-through all data while incurring minimal overhead.
This comment was in direct reference to this commit:

```I'm not sure how people feel about commit 5975019; it adds in the functionality to encode and decode directly to/from bytes without having a connection open. I found this functionality useful, but it might be too use-case specific.```

Additionally, you can turn a byte slice into a bufio.Reader using:  ```bufio.NewReader(bytes.NewReader(byteSlice))```